### DISCUSSION TITLE
try-with-resources 에 대해서 설명해주세요.

### DISCUSSION THREAD

**Discussion by le2sky** [2024-woowacourse-study/level-interview#51] [Java]
그리고 왜 나왔을까요!?

**Comment by jimi567** [comment]
try-with-resources 구문은 AutoCloseable 인터페이스가 구현된 클래스의 자원을 자동으로 해제하는 데 사용된다. 주로 파일이나 데이터베이스 연결과 같은 자원을 사용하는 경우에 많이 활용된다. 

또한,  try () 괄호 안에 여러 개의 리소스를 선언하고 초기화할 수 있고, 예외가 발생하면 catch 블록으로 제어가 전달되는데, 이때 리소스가 안전하게 해제된다.
try-with-resources를 사용하면, 자원 관리를 보다 간편하게 할 수 있고, 코드 안정성과 가독성을 향상시킬 수 있다.

**Comment by nayonsoso** [comment]
외부의 리소스는 사용하면 닫아주는 것이 중요합니다.
왜냐하면 다른 곳에서도 사용할 수 있기 때문입니다.

이처럼 사용 후, 반납해야 하는 자원을 try-with-resource로 관리할 수 있습니다.
try-with-resource는 autoCloeable 인터페이스를 구현하는 리소스를 자동으로 닫아주는 역할을 합니다.
try 블럭을 벗어나는 순간 자동으로 `close()` 메서드가 호출되는 방식으로 동작합니다.

try-with-resource가 없던 시절에는 직접 자원을 닫아줘야 했는데요,
이런 방식은 아래의 단점이 있었습니다.
1. 코드가 복잡해집니다.
2.  close 구문을 누락할 여지가 있습니다. 
3. 예외가 발생할 경우 자원 반납을 하지 못 할 수 있습니다.

try-with-resource는 이런 문제들을 해결해줄 수 있는 좋은 수단입니다.

**Comment by le2sky** [comment]
https://tecoble.techcourse.co.kr/post/2021-04-26-try-with-resource/

**Comment by GIVEN53** [comment]
자원 자동 반납
java 7 : try-catch-finally -> 휴먼 에러, 에러에 의해 누락, 가독성 감소
조건: AutoClosable을 구현한 구현체, try 괄호 내에서 리소스 호출
특징, 자동 close() 호출, 나중에 선언한 리소스부터 반납
